File: api/src/new_phone/auth/jwt.py
import uuid
from datetime import UTC, datetime, timedelta

from jose import jwt

from new_phone.config import settings


def create_access_token(
    user_id: uuid.UUID,
    tenant_id: uuid.UUID,
    role: str,
    language: str = "en",
) -> str:
    expire = datetime.now(UTC) + timedelta(minutes=settings.jwt_access_token_expire_minutes)
    payload = {
        "sub": str(user_id),
        "tenant_id": str(tenant_id),
        "role": role,
        "language": language,
        "type": "access",
        "exp": expire,
        "iat": datetime.now(UTC),
    }
    return jwt.encode(payload, settings.jwt_secret_key, algorithm=settings.jwt_algorithm)


def create_refresh_token(user_id: uuid.UUID) -> tuple[str, datetime]:
    expire = datetime.now(UTC) + timedelta(days=settings.jwt_refresh_token_expire_days)
    payload = {
        "sub": str(user_id),
        "type": "refresh",
        "exp": expire,
        "iat": datetime.now(UTC),
        "jti": str(uuid.uuid4()),
    }
    token = jwt.encode(payload, settings.jwt_secret_key, algorithm=settings.jwt_algorithm)
    return token, expire


def create_mfa_token(user_id: uuid.UUID) -> str:
    """Short-lived token for MFA challenge step."""
    expire = datetime.now(UTC) + timedelta(minutes=5)
    payload = {
        "sub": str(user_id),
        "type": "mfa_pending",
        "exp": expire,
        "iat": datetime.now(UTC),
    }
    return jwt.encode(payload, settings.jwt_secret_key, algorithm=settings.jwt_algorithm)


def decode_token(token: str) -> dict:
    """Decode and validate a JWT. Raises JWTError on failure."""
    return jwt.decode(token, settings.jwt_secret_key, algorithms=[settings.jwt_algorithm])

File: web/src/lib/jwt.ts
export interface JwtPayload {
  sub: string       // user_id UUID
  tenant_id: string // UUID
  role: "msp_super_admin" | "msp_tech" | "tenant_admin" | "tenant_manager" | "tenant_user"
  type: "access" | "refresh"
  language?: string
  exp: number
  iat: number
}

export function decodeJwtPayload(token: string): JwtPayload {
  const parts = token.split(".")
  if (parts.length !== 3) throw new Error("Invalid JWT")
  const payload = parts[1]
  const decoded = atob(payload.replace(/-/g, "+").replace(/_/g, "/"))
  return JSON.parse(decoded)
}

export function isTokenExpired(token: string, bufferSeconds = 30): boolean {
  try {
    const payload = decodeJwtPayload(token)
    return Date.now() / 1000 > payload.exp - bufferSeconds
  } catch {
    return true
  }
}
